File: strategy/cpp/strategy_test.cc
#include <gtest/gtest.h>
#include "strategy/cpp/strategy.h"
#include <string>

TEST(PaymentTest, CashPay) {
    Payment p("alice", "bob", 100, std::make_shared<CacheStrategy>());

    std::string out = p.Pay();
    std::string want = "Pay $100 to bob by cash";

    EXPECT_EQ(out, want);
}

TEST(PaymentTest, BankPay) {
    Payment p("alice-bank", "bob", 200, std::make_shared<BanckStrategy>());

    std::string out = p.Pay();
    std::string want = "Pay $200 to bob by bank account alice-bank";

    EXPECT_EQ(out, want);
}

File: strategy/strategy_test.go
package strategy

import (
	"testing"
)

func TestCashPay(t *testing.T) {
	p := NewPayment("alice", "bob", 100, &Cash{})

	want := "Pay $100 to alice by cash"
	out := p.Pay()

	if out != want {
		t.Fatalf("expected %q, got %q", want, out)
	}
}

func TestBankPay(t *testing.T) {
	p := NewPayment("alice-bank", "bob", 200, &Bank{})

	out := p.Pay()

	want := "Pay $200 to alice-bank by bank account bob"

	if out != want {
		t.Fatalf("expected %q, got %q", want, out)
	}
}
